### ISSUE TITLE
Brainstorm your ideas for optimization of Kratos here

### ISSUE THREAD

**Issue by marandra** [KratosMultiphysics/Kratos#9896]
I am opening this issue to collect ideas for improving the performance of Kratos, making use of the collective knowledge and the variety of backgrounds of its great community. 

Ideas may be, e.g., something to explore (especially if its implementation is easy), solutions already implemented that you consider are worth implementing in general, technological advances that we are missing. They may cover different aspects, such as optimizations at compile time, parallelization, linear solvers, architecture (CPU, GPU), numerical methods (predictors, preconditioners), physical models (elements, constitutive laws), “recipes” (combination of parameters, optimal settings).

My idea is to keep track of the progress of the individual actions implemented, as well as the evolution of the performance between main releases of Kratos. For this, I would like to write or repurpose some existing tests into a benchmark pool, so we can have an objective measurement of the progress and to have it documented, for example, on a wiki page. It would be nice also to have tips, best practices, recipes, configurations, etc.

I will be working on some of the items myself, mainly the ones related to the projects I am currently working on. Regardless, I'd like to allocate some time for other needed actions like planning, writing, consulting, delegating, etc., so it is easier to keep track of the work of other people.

To sum up: My intent is that this is a place to share current or future ideas for improving the performance of Kratos. I will be working on some of them, as well as keeping track of the implementation of these ideas, their conclusions and, hopefully, the improvements.

Feel free to share your thoughts. Thank you in advance.

**Comment by marandra** [comment]
This is a list of things I am already considering or working on. Feel free to edit, add, or update this message. Ideas, suggestions and contributions are very welcome.

Low level (mainly profiling): 
- Automatic optimization depending on the compiler and its options, attributes, etc.
- Automatic vectorization of loops (i.e. -favx2). Not easy given the OO nature of Kratos
- Memory bottlenecks, analysis with roofline models.
- Behavior of shared memory parallelization (looks ok) 

Core:
- Linear solver (e.g., specialized solvers, configuration of AMGCL, running on GPUs)

Fluid Dynamics Application:
- Element coded using symbolic math (WIP)
- Unroll gauss points loops
- Explore predictors to reduce the number of nonlinear iterations (WIP)

**Comment by loumalouomega** [comment]
FYI: https://www.agner.org/optimize/optimizing_cpp.pdf

**Comment by ddiezrod** [comment]
Thanks for opening this! We did some investigation in the past and at least in the context of fluid dynamics the two biggest bottlenecks were:
- The assembly of the elements
- The use of ilu0 in smoother steps in AMGCL. ilu0 is pretty much serial, but we could not get rid of it in our flow solver in order to have a robust solution.

ping @jrubio

**Comment by RiccardoRossi** [comment]
@ddiezrod this data are with gcc or with msvc?

**Comment by ddiezrod** [comment]
> @ddiezrod this data are with gcc or with msvc?

only msvc

**Comment by RiccardoRossi** [comment]
well, msvc optimizer is not really known for its efficiency...

**Comment by EduardGomezEscandell** [comment]
Once we move to c++17 we may be able to make many functions `constexpr`. In c++11 there can only be one statement in a constexpr function, which is quite restrictive. In c++17 this limitation is lifted. I don't foresee big performance improvements, since most of our computation is unknown at compile-time, but it should help the optimizer by letting it have more info at compile-time in some cases.

If we really wanted to: quadrature rules, shape functions, and their gradient (in isoparametric form) are known beforehand and we may be able to make constexpr. This should help the compiler at optimizing element assembly.

All of this at the cost of larger binaries though.

**Comment by AlejandroCornejo** [comment]
I have been working in the optimization of the product Bt*D*B in solid elements inside the SMA. 

I have noticed that a benchmark of calling the `BuildLocalSystem ` of **1M elements in twice faster if you compile in Linux with respect to the Windows compilation**. This is quite surprising...

**Comment by loumalouomega** [comment]
> I have been working in the optimization of the product Bt_D_B in solid elements inside the SMA.
> 
> I have noticed that a benchmark of calling the `BuildLocalSystem ` of **1M elements in twice faster if you compile in Linux with respect to the Windows compilation**. This is quite surprising...

From the document I sent:

> Microsoft Visual Studio
> This is a very user friendly compiler with many features and a good debugger. The full version is very expensive, but a limited non-commercial version is available for free. Visual Studio can build code with various programming languages (C++, C#, Visual Basic) for various platforms. Visual Studio is not the best compiler in terms of code optimization, but it can be very useful during the development phase of a project. You may use a different compiler for the release version of your code if performance is important. It is possible to  use the Visual Studio integrated development environment (IDE) with a different compiler.

You can try to compile with Gcc in Windows with MinGW:

https://github.com/KratosMultiphysics/Kratos/wiki/Windows-Install-(MinGW)

**TODO**: Copy this in the INSTALL.md

**Comment by miguelmaso** [comment]
> My idea is to keep track of the progress of the individual actions implemented, as well as the evolution of the performance between main releases of Kratos. For this, I would like to write or repurpose some existing tests into a benchmark pool, so we can have an objective measurement of the progress and to have it documented, for example, on a wiki page. It would be nice also to have tips, best practices, recipes, configurations, etc.

It would be interesting to add a CI to the examples repo. Probably, using pip. Apart from being an indirect testing of the release, some global results can be tracked (e.g, RMS of error, time consumption)

**Comment by Vahid-Galavi** [comment]
Hi,

Good that this issue is opened. I have already opened issue #9890. According to my experience, build time is not a bottleneck and it is similar to the build time in some commercial software. Time is usually spent on solving the system. Therefore, the linear solvers should be optimised. Maybe, we should implement parallel linear solvers based on the domain decomposition technique.

For the systems, in which the stiffness matrix does not change during the iterations, I think we do not need to solve the stiffness matrix per iteration over and over. Maybe, we need to store the decomposed matrix and reused it.

**Comment by AlejandroCornejo** [comment]
> 

You mean like the AMGCL? Have you tried pardiso? :)

**Comment by Vahid-Galavi** [comment]
@AlejandroCornejo I used AMGCL. I couldn't get Pardiso running.

**Comment by ddiezrod** [comment]
> @AlejandroCornejo I installed it on Linux. It returned an internal error during decomposition. I will try it on Windows to see if it helps.

We had problems using pardiso in linux when using python versions lower than 3.8. Something get messed up that makes the decomposition fail. We decided to update to python 3.8 just to solve this problem. It might be the same problem you had.

**Comment by Vahid-Galavi** [comment]
@ddiezrod I compared it with PLAXIS, which is quite fast. Build time is similar, however, PLAXIS decomposes the stiffness matrix only once (if the stiffness matrix is constant) and reuses it in the subsequent iterations/steps. Kratos does not have such a feature. This led to a huge difference in computation time (~mins in PLAXIS vs hours in KRATOS!). In this example, the build time is around 0.2 s while the solve time is between 10-20 s.

**Comment by matekelemen** [comment]
Parts of my pet project saw a significant (~5%-20%) performance hit when I switched to compiling with position independent code ```-fPIC``` enabled. I know we need position independence because of *pybind*, but I'm wondering whether it's possible to partition our build such that not all binaries require ```-fPIC```, or at least that we can replace ```-fPIC``` with ```-fPIE```.

**Comment by AlejandroCornejo** [comment]
> > I have been working in the optimization of the product Bt_D_B in solid elements inside the SMA.
> > I have noticed that a benchmark of calling the `BuildLocalSystem ` of **1M elements in twice faster if you compile in Linux with respect to the Windows compilation**. This is quite surprising...
> 
> From the document I sent:
> 
> > Microsoft Visual Studio
> > This is a very user friendly compiler with many features and a good debugger. The full version is very expensive, but a limited non-commercial version is available for free. Visual Studio can build code with various programming languages (C++, C#, Visual Basic) for various platforms. Visual Studio is not the best compiler in terms of code optimization, but it can be very useful during the development phase of a project. You may use a different compiler for the release version of your code if performance is important. It is possible to  use the Visual Studio integrated development environment (IDE) with a different compiler.
> 
> You can try to compile with Gcc in Windows with MinGW:
> 
> https://github.com/KratosMultiphysics/Kratos/wiki/Windows-Install-(MinGW)
> 
> **TODO**: Copy this in the INSTALL.md

Following this issue, I implemented a hardcoded version of the BtDB product (by using sympy and automatic differentiation). and my test in **Linux has reduced its execution time a 34%**. In Windows the execution time *increased* a 100% (almost twice)...

**Comment by loumalouomega** [comment]
> Parts of my pet project saw a significant (~5%-20%) performance hit when I switched to compiling with position independent code `-fPIC` enabled. I know we need position independence because of _pybind_, but I'm wondering whether it's possible to partition our build such that not all binaries require `-fPIC`, or at least that we can replace `-fPIC` with `-fPIE`.

Usually when libraries are compiled static the code can be faster. In fact, many performance issues from Windows and Mac come from their respective linking strategies

**Comment by matekelemen** [comment]
> Usually when libraries are compiled static the code can be faster. In fact, many performance issues from Windows and Mac come from their respective linking strategies

Right, so would there be any benefit in building *Kratos* libraries as static ones, or would it make no difference because we're linking them to *pybind* modules that are shared anyway?

**Comment by loumalouomega** [comment]
> Right, so would there be any benefit in building _Kratos_ libraries as static ones, or would it make no difference because we're linking them to _pybind_ modules that are shared anyway?

I did a small research and a priori can be done, but not recommended.

https://github.com/pybind/pybind11/issues/466

Any case apparently with fpic can be improved. In any case, we can compile other things as static, like the solvers. It could improve performance.

**Comment by mpentek** [comment]
I propose moving these inputs from "Issues" to "Discussions". 

Also, I think code profiling and analyzing would need to be addressed. We are currently trying out VTune and Maqao again on Kratos. Will post some updates and actualize the HoTos.

The most difficult aspect will be to go from "Brainstorming->Ideas" to actionable tasks. Otherwise the time can be better invested.
